### ISSUE TITLE
cuSOLVER backend for Sparse CSR direct linear solvers

### ISSUE THREAD

**Issue by IvanYashchuk** [pytorch/pytorch#69538]
## 🚀 Feature

cuSolverSP (part of cuSOLVER) provides linear solver, least-square solver routines for matrices in CSR format. We can use it as a backend for `torch.linalg.solve` and `torch.linalg.lstsq`.

We are working on improving coverage of Sparse CSR matrices in PyTorch. Solving systems of linear equations is one of the fundamental operations in computational linear algebra.

From cuSOLVER docs:
> The high-level API is designed for ease-of-use, so it allocates any required memory under the hood automatically. If the host or GPU system memory is not enough, an error is returned.

Unfortunately, it means these functions do not allow passing externally allocated buffers, and paths with inputs on GPU are not implemented (there remarks for relevant functions "only CPU (Host) path is provided").

- [ ] `torch.linalg.solve`
    Use [`cusolverSp<t>csrlsvlu`](https://docs.nvidia.com/cuda/cusolver/index.html#cusolver-lt-t-gt-csrlsvlu). 

- [ ] `torch.linalg.lstsq`
    Use [`cusolverSp<t>csrlsqvqr`](https://docs.nvidia.com/cuda/cusolver/index.html#cusolver-lt-t-gt-csrlsqvqr)

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk @ngimel @jianyuh @mruberry @walterddr @xwang233 @Lezcano

**Comment by Lezcano** [comment]
the lstsq routine is not in a fantastic shape given its crazy number of returns. We might want to wait to implement `solve_lstsq` (or `solve_nonsquare` perhaps?) which would be simply `lstsq` with a nicer API (i.e. no driver keyword, but rather some more intuitive API via a more descriptive paramenter (cf. the note explaining these parameters [in the docs](https://pytorch.org/docs/stable/generated/torch.linalg.lstsq.html)). Within this nice API, it would also be cool to give support to the different backends on CUDA via `linalg.svd` or so.

What I'm trying to say here is that we could perhaps postpone that task if there's other tasks that need doing first, so that it does not happen like with `torch.triangular_solve`.

**Comment by IvanYashchuk** [comment]
On CUDA `torch.linalg.lstsq` supports only one "driver" that is a QR-based solver and that's also the only solver that is available in cuSOLVER for sparse matrices. It would be easy to port the code to any new future frontend of `lstsq` functionality.

**Comment by Lezcano** [comment]
If you think that then porting the code to whatever's the next operation would be easy (it makes sense) then I have no complaints.

**Comment by tvercaut** [comment]
I am interested in sparse linear solvers as well. As a current workaround, I tried using cupy as a backend since their [`spsolve`](https://docs.cupy.dev/en/stable/reference/generated/cupyx.scipy.sparse.linalg.spsolve.html) function wraps [the cusolver `csrlsvqr`](https://docs.nvidia.com/cuda/cusolver/index.html#cusolver-lt-t-gt-csrlsvqr):
https://github.com/cupy/cupy/blob/fddaf5e901930512d97bd9ecfd4e0e5dfe0fc56b/cupyx/scipy/sparse/linalg/_solve.py#L492

Unfortunately I realised that `csrlsvqr` and related direct solvers in cusolver only seem to support vector righ-hand sides:
https://docs.nvidia.com/cuda/cusolver/index.html#cusolver-highlevel-function-reference
and thus `spsolve` in cupy only handle vector righ-hand sides as well.

As a further fallback, I thus tried using the LU factorisation in cupy but this factorisation doesn't take advantage of the GPU. Only the solving, post factorisation, runs on CUDA:
https://docs.cupy.dev/en/stable/reference/generated/cupyx.scipy.sparse.linalg.factorized.html

For what it's worth, my current workaround (without any batch support) based on [this blog post](https://blog.flaport.net/solving-sparse-linear-systems-in-pytorch.html) reads as such (full working notebook [here](https://colab.research.google.com/gist/tvercaut/66fe2fc73b56592cbbf48d86ba0fa3e0/sparseinverse-torchcupy-interop-test.ipynb)):
```python
# Convenience function to map a torch COO tensor in a cupy one
def coo_torch2cupy(A):
  A = A.data.coalesce()
  Avals_cp = cp.asarray(A.values())
  Aidx_cp = cp.asarray(A.indices())
  return cp.sparse.coo_matrix((Avals_cp, Aidx_cp))

# Custom PyTorch sparse solver exploiting a CuPy backend
# See https://blog.flaport.net/solving-sparse-linear-systems-in-pytorch.html
class SparseSolve(torch.autograd.Function):
  @staticmethod
  def forward(ctx, A, b):
    # Sanity check
    if A.ndim != 2 or (A.shape[0] != A.shape[1]):
      raise ValueError("A should be a square 2D matrix.")
    # Transfer data to CuPy
    A_cp = coo_torch2cupy(A)
    b_cp = cp.asarray(b.data)
    # Solver the sparse system
    ctx.factorisedsolver = None
    if (b.ndim == 1) or (b.shape[1] == 1):
      # cp.sparse.linalg.spsolve only works if b is a vector but is fully on GPU
      x_cp = cp.sparse.linalg.spsolve(A_cp, b_cp)
    else:
      # Make use of a factorisation (only the solver is then on the GPU)
      # We store it in ctx to reuse it in the backward pass
      ctx.factorisedsolver = cp.sparse.linalg.factorized(A_cp)
      x_cp = ctx.factorisedsolver(b_cp)
    # Transfer (dense) result back to PyTorch
    x = torch.as_tensor(x_cp, device=b.device)
    if A.requires_grad or b.requires_grad:
      # Not sure if the following is needed / helpful
      x.requires_grad = True
    else:
      # Free up memory
      ctx.factorisedsolver = None
    # Save context for backward pass
    ctx.save_for_backward(A, b, x)
    return x

  @staticmethod
  def backward(ctx, grad):
    # Recover context
    A, b, x = ctx.saved_tensors
    # Compute gradient with respect to b
    if (ctx.factorisedsolver is None):
      gradb = SparseSolve.apply(A.t(), grad)
    else:
      # Re-use factorised solver from forward pass
      grad_cp = cp.asarray(grad.data)
      gradb_cp = ctx.factorisedsolver(grad_cp, trans='T')
      gradb = torch.as_tensor(gradb_cp, device=b.device)
    # The gradient with respect to the (dense) matrix A would be something like
    # -gradb @ x.T but we are only interested in the gradient with respect to
    # the (non-zero) values of A
    gradAidx = A.indices()
    mgradbselect = -gradb.index_select(0,gradAidx[0,:])
    xselect = x.index_select(0,gradAidx[1,:])
    mgbx = mgradbselect * xselect
    if x.dim() == 1:
      gradAvals = mgbx
    else:
      gradAvals = torch.sum( mgbx, dim=1 )
    gradAs = torch.sparse_coo_tensor(gradAidx, gradAvals, A.shape)
    return gradAs, gradb

sparse_solve = SparseSolve.apply
```

It would be gread to be able to scrap this workaround!

**Comment by IvanYashchuk** [comment]
Thank you, @tvercaut, for posting your `SparseSolve` function!

@krshrimali is currently working on getting `csrlsvqr` integrated into PyTorch, so the workaround won't be necessary at some point in the future.

It has to be discussed separately but it's unlikely that PyTorch would introduce a dependency on SuperLU or SuiteSparse' KLU for sparse LU factorization due to maintenance and licensing issues.

**Comment by tvercaut** [comment]
Sounds great. I had a quick check but there doesn't seem to be a simple means of using cuSolver to first factorise the sparse matrix and then solve the linear system. At least not with any existing python bindings I could try. That said, I saw this approach which uses [cuSolverRF](https://docs.nvidia.com/cuda/cusolver/index.html#cuSolverRF-reference) to achieve that:
https://github.com/zishun/cuSolverRf-batch

I guess something like this could be an alternative to `csrlsvqr` that would work by multiple right-hand sides (i.e. b being a matrix) and would also allow saving the factorisation for reuse in the backward pass.

**Comment by tvercaut** [comment]
Thanks @Lezcano. Indeed, the decomposition may not be sparse but my understanding was that cuSOLVER and similar direct sparse solver maintained an implicit memory-efficient representation of the decomposition suitable for solving purposes which thus make them often useful in practice.
